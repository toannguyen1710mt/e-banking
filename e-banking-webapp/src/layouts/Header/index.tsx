'use client';

// Libs
import {
  Avatar,
  NavbarBrand,
  NavbarContent,
  NavbarMenuToggle,
  Navbar as NavbarNextUI,
  Spinner,
} from '@nextui-org/react';
import { useState } from 'react';
import Image from 'next/image';
import Link from 'next/link';
import { Session } from 'next-auth';
import { redirect, usePathname } from 'next/navigation';

// Constants
import { ERROR_MESSAGES, IMAGES, NavbarItem, REGEX, ROUTES } from '@/constants';

// Components
import {
  Button,
  MenuDropdown,
  Navbar,
  NotificationIcon,
  SearchIcon,
  SettingIcon,
  SignOutIcon,
} from '@/components';

// Actions
import { signOut } from '@/actions';

// Contexts
import { useToastContext } from '@/context';

interface IHeaderProps {
  session: Session;
}

export const Header = ({ session }: IHeaderProps) => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const [isSigningOut, setIsSigningOut] = useState(false);
  const { showToast } = useToastContext();
  const pathname = usePathname();

  const isSettingsUrl = REGEX.SETTINGS.test(pathname);

  const {
    user: { username, email, avatar },
  } = session;

  const mobileMenuOptions = [
    ...NavbarItem.map((item) => ({
      key: item.path,
      label: item.text,
    })),
  ];

  const handleSignOut = async () => {
    setIsSigningOut(true);

    try {
      await signOut();

      showToast(ERROR_MESSAGES.SIGN_OUT_SUCCESS, 'success', 'top-center');
    } catch (error) {
      console.error(error);

      showToast(ERROR_MESSAGES.SIGN_OUT_FAILED, 'error', 'top-center');
    } finally {
      setIsSigningOut(false);

      redirect(ROUTES.SIGN_IN);
    }
  };

  const dropDownMenuOptions = [
    {
      key: username,
      label: '',
      customOptionElement: (
        <div className='flex flex-col gap-2'>
          <span className='text-xs font-semibold'>{username}</span>
          <span className='text-3xs font-semibold text-transparentBlack'>
            {email}
          </span>
        </div>
      ),
    },
    {
      key: 'Settings',
      label: 'Settings',
      startContent: <SettingIcon />,
      customOptionElement: <Link href={ROUTES.GENERAL}>Settings</Link>,
    },
    {
      key: 'Sign Out',
      label: 'Sign Out',
      customOptionElement: (
        <Button
          startContent={!isSigningOut && <SignOutIcon />}
          spinner={<Spinner size='sm' />}
          isLoading={isSigningOut}
          isDisabled={isSigningOut}
          onClick={handleSignOut}
          className='flex h-full !max-h-none min-w-0 justify-start gap-2 !bg-transparent p-0 text-foreground-100'
        >
          Sign out
        </Button>
      ),
    },
  ];

  return (
    <NavbarNextUI
      classNames={{
        wrapper: `max-w-full justify-start border-b border-foreground-100 border-opacity-25 px-3 ${isSettingsUrl ? 'bg-background-500' : 'bg-background-900'}`,
      }}
      onMenuOpenChange={setIsMenuOpen}
    >
      <div className='flex items-center gap-[88px]'>
        <NavbarBrand className='gap-5'>
          <MenuDropdown
            options={mobileMenuOptions}
            customTriggerElement={
              <NavbarMenuToggle
                aria-label={isMenuOpen ? 'Close menu' : 'Open menu'}
                className='hidden max-[700px]:flex'
              />
            }
            isDivided={true}
          />
          <Link href={ROUTES.HOME}>
            <Image
              src={IMAGES.LOGO}
              alt='Logo EBanking'
              width={104}
              height={36}
            />
          </Link>
        </NavbarBrand>
        <NavbarContent
          className='flex gap-[59px] font-normal text-transparentBlack max-[700px]:hidden'
          data-justify='center'
        >
          <Navbar navbarItem={NavbarItem} />
        </NavbarContent>
      </div>
      <NavbarContent data-justify='end'>
        <button
          aria-label='search icon'
          disabled={true}
          className='cursor-not-allowed'
        >
          <SearchIcon />
        </button>
        <button
          aria-label='notification icon'
          disabled={true}
          className='cursor-not-allowed'
        >
          <NotificationIcon />
        </button>
        <MenuDropdown
          customTriggerElement={
            <Avatar src={avatar} className='h-6 w-6 rounded-full' />
          }
          options={dropDownMenuOptions}
          isDivided={true}
        />
      </NavbarContent>
    </NavbarNextUI>
  );
};
