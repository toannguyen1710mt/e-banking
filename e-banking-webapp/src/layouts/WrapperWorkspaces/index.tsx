'use client';

import { usePathname } from 'next/navigation';

// Constants
import { REGEX } from '@/constants';

export const WrapperWorkspaces = ({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) => {
  const pathname = usePathname();

  const isSettingsUrl = REGEX.SETTINGS.test(pathname);

  return (
    <div
      className={`${isSettingsUrl ? 'bg-background-500 p-0' : 'bg-background-900 p-3'} flex grow`}
    >
      {children}
    </div>
  );
};
