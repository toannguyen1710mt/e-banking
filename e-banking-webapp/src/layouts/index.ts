export * from './Header';
export * from './WrapperWorkspaces';
