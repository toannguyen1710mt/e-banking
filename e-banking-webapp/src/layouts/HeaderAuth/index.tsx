// Configs
import { auth } from '@/config/auth';

// Layouts
import { Header } from '@/layouts';

export const HeaderAuth = async () => {
  const session = await auth();

  if (!session) return null;

  return <Header session={session} />;
};
