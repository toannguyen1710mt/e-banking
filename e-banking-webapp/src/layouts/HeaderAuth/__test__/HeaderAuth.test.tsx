import { HeaderAuth } from '@/layouts';

describe('HeaderAuth', () => {
  it('should return undefined if there is no session', async () => {
    const authMock = jest.fn(() => Promise.resolve(null));

    authMock.mockResolvedValueOnce(null);

    const result = await HeaderAuth();

    expect(result).toMatchSnapshot();
  });
});
