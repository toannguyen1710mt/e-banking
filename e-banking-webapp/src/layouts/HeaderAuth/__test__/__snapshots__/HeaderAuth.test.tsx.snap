// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderAuth should return undefined if there is no session 1`] = `null`;
